src/coreclr/vm/codeman.cpp

```diff
@@ -3917,126 +3984,380 @@ StubCodeBlockKind EEJitManager::GetStubCodeBlockKind(RangeSection * pRangeSectio
+        //     1. Look up DWORD representing given PC
+        //     2. If DWORD is a Pointer, then we are done. Otherwise,
+        //     3. If nibble corresponding to PC is initialized, if the value the nibble represents precedes the PC return that value, otherwise
+        //     4. Find the next preceding initialized nibble in the DWORD. If found, return the value the nibble represents. Otherwise,
+        //     5. Execute steps 1, 2 then 4 on the proceeding DWORD. If this is also uninitialized, then we are not in a function and can return a nullptr.
+        static TADDR FindMethodCode(RangeSection * pRangeSection, PCODE currentPC)
+        {
+            LIMITED_METHOD_DAC_CONTRACT;
 
-    if (startPos < NIBBLES_PER_DWORD)
-        return 0;
+            _ASSERTE(pRangeSection != NULL);
+            _ASSERTE(pRangeSection->_flags & RangeSection::RANGE_SECTION_CODEHEAP);
 
-    startPos = ((startPos >> LOG2_NIBBLES_PER_DWORD) << LOG2_NIBBLES_PER_DWORD) - 1;
+            // remove bottom two bits to ensure alignment math
+            // on ARM32 Thumb, the low bits indicate the thumb instruction set
+            currentPC = ALIGN_DOWN(currentPC, CODE_ALIGN);
 
-    // Skip "headerless" DWORDS
+            HeapList *pHp = pRangeSection->_pHeapList;
 
-    while (pMapStart < pMap && 0 == (tmp = VolatileLoadWithoutBarrier<DWORD>(--pMap)))
-    {
-        startPos -= NIBBLES_PER_DWORD;
-    }
+            if ((currentPC < pHp->startAddress) ||
+                (currentPC > pHp->endAddress))
+            {
+                return 0;
+            }
 
-    // This helps to catch degenerate error cases. This relies on the fact that
-    // startPos cannot ever be bigger than MAX_UINT
-    if (((INT_PTR)startPos) < 0)
-        return 0;
+            TADDR base = pHp->mapBase;
+            TADDR delta = currentPC - base;
 
-    // Find the nibble with the header in the DWORD
+            size_t dwordIndex = GetDwordIndex(delta);
+            size_t nibbleIndex = GetNibbleIndex(delta);
 
-    while (startPos && !(tmp & NIBBLE_MASK))
-    {
-        tmp = tmp >> NIBBLE_SIZE;
-        startPos--;
-    }
+            PTR_DWORD pMap = pHp->pHdrMap;
 
-    if (startPos == 0 && tmp == 0)
-        return 0;
+            DWORD dword;
+            DWORD nibble;
+
+            DWORD  offset   = ADDR2OFFS(delta); // this is the offset inside the bucket + 1
+            _ASSERTE(offset == (offset & NIBBLE_MASK));
+
+            // #1 look up DWORD represnting current PC
+            PREFIX_ASSUME(pMap + dwordIndex != NULL);
+            dword = VolatileLoadWithoutBarrier<DWORD>(pMap + dwordIndex);
+
+            // if DWORD is empty, we can skip steps 2,3, and 4
+            if (dword)
+            {
+                // #2 if DWORD is a pointer, then we can return
+                if (IsPointer(dword))
+                {
+                    TADDR newAddr = base + DecodePointer(dword);
+                    return newAddr;
+                }
```

Comment:
if the dword is a pointer, shoudln't you then do a second read at the code start to validate that the method isn't being deleted at the same time?

Reply:
Similar to above. I agree that there could be some safety issues if concurrent delete/reads occur. 

However, I do not think this changes the previous behavior. The current nibble map allows concurrent delete/reads and does not verify that the function is still valid when reading.

My understanding (may be wrong) is that the code that calls write and delete must be careful about when methods can be removed from the nibblemap.